### PR TITLE
WFLY-8315 JGroups protocols collect statistics by default

### PR DESCRIPTION
**Pull request by rhusar**
Jiras
https://issues.jboss.org/browse/WFLY-8315
https://issues.jboss.org/browse/WFLY-2453
https://issues.jboss.org/browse/WFLY-8317

### CODE REVIEW

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/ForkChannelFactory.java [hunk 1]
```diff
@@ -58,7 +58,11 @@ public Channel createChannel(String id) throws Exception {
 
         String stackName = this.protocols.isEmpty() ? this.channel.getClusterName() : id;
 
-        return new ForkChannel(this.channel, stackName, id, this.protocols.stream().map(ProtocolConfiguration::createProtocol).toArray(Protocol[]::new));
+        return new ForkChannel(this.channel, stackName, id, this.protocols.stream().map(protocolConfiguration -> {
+            Protocol protocol = protocolConfiguration.createProtocol();
+            protocol.enableStats(protocolConfiguration.isStatisticsEnabled() != null ? protocolConfiguration.isStatisticsEnabled() : this.parentFactory.getProtocolStackConfiguration().isStatisticsEnabled());
```
Comment (pferraro) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/ForkChannelFactory.java]: Why not a separate .map(...) function?

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java [hunk 2]
```diff
@@ -64,16 +66,19 @@ public Channel createChannel(String id) throws Exception {
...
         ProtocolStack stack = new ProtocolStack();
         channel.setProtocolStack(stack);
-        stack.addProtocol(this.configuration.getTransport().createProtocol());
-        stack.addProtocols(this.configuration.getProtocols().stream().map(ProtocolConfiguration::createProtocol).collect(Collectors.toList()));
-
-        RelayConfiguration relay = this.configuration.getRelay();
-        if (relay != null) {
-            stack.addProtocol(relay.createProtocol());
-        }
+        stack.addProtocols(Stream.of(
```
Comment (pferraro) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java]: Stream.concat(...) would be much more efficient here.

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java [hunk 3]
```diff
@@ -64,16 +66,19 @@ public Channel createChannel(String id) throws Exception {
...
+                Stream.of(this.configuration.getTransport()),
+                this.configuration.getProtocols().stream(),
+                Stream.of(this.configuration.getRelay())
+        ).flatMap(Function.identity()).filter(Objects::nonNull).map(protocolConfiguration -> {
+            Protocol protocol = protocolConfiguration.createProtocol();
+            protocol.enableStats(protocolConfiguration.isStatisticsEnabled() != null ? protocolConfiguration.isStatisticsEnabled() : this.configuration.isStatisticsEnabled());
```
Comment (pferraro) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java]: .map(protocol -> protocol.enableStats(...)) ?

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java [hunk 4]
```diff
@@ -64,16 +66,19 @@ public Channel createChannel(String id) throws Exception {
...
+            return protocol;
+        }).collect(Collectors.toList()));
 
         // Add implicit FORK to the top of the stack
         FORK fork = new FORK();
+        fork.enableStats(configuration.isStatisticsEnabled());
```
Comment (pferraro) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java]: Probably easier to add () -> new FORK() to the stream created above.

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/AbstractProtocolConfigurationBuilder.java [hunk 5]
```diff
@@ -80,6 +81,7 @@ protected AbstractProtocolConfigurationBuilder(String name) {
     public Builder<C> configure(OperationContext context, ModelNode model) throws OperationFailedException {
         this.moduleName = MODULE.resolveModelAttribute(context, model).asString();
         this.properties = ModelNodes.optionalPropertyList(PROPERTIES.resolveModelAttribute(context, model)).orElse(Collections.emptyList()).stream().collect(Collectors.toMap(Property::getName, property -> property.getValue().asString()));
+        this.statisticsEnabled = STATISTICS_ENABLED.resolveModelAttribute(context, model).asBooleanOrNull();
```
Comment (pferraro) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/AbstractProtocolConfigurationBuilder.java]: Nice.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/jgroups/spi/src/main/java/org/wildfly/clustering/jgroups/spi/ProtocolConfiguration.java [hunk 6]
```diff
@@ -31,5 +33,20 @@
...
+    /**
+     * Whether or not this protocol will collect statistics overriding stack configuration.
+     * @return {@code true}/{@code false}, whether to enable/disable statistics for the protocol;
+     *         {@code null} if the stack configuration will be honored
+     */
+    Boolean isStatisticsEnabled();
+
     P createProtocol();
+
+    static Function<ProtocolConfiguration<? extends Protocol>, Protocol> creator(ProtocolStackConfiguration stackConfiguration) {
```
Comment (pferraro) [inline clustering/jgroups/spi/src/main/java/org/wildfly/clustering/jgroups/spi/ProtocolConfiguration.java]: I just vomited in my mouth a little...

Comment (pferraro) [inline clustering/jgroups/spi/src/main/java/org/wildfly/clustering/jgroups/spi/ProtocolConfiguration.java]: :)

**Review by pferraro — APPROVED** [review]

**Comment by rhusar** [comment]
retest this please

**Comment by kabir** [comment]
Retest this please

**Pull request merged**